Plugins/GeoModel/src/GeoModelToDetVol.cpp

```diff
@@ -0,0 +1,182 @@
+                                               tube.getZHalfLength());
+    volumes.emplace_back(Experimental::DetectorVolumeFactory::construct(
+        portalGenerator, context, name, transform, bounds,
+        Experimental::tryAllPortalsAndSurfaces()));
+  } else if (shape.typeID() == GeoTubs::getClassTypeID()) {
+    const GeoTubs& tubs = static_cast<const GeoTubs&>(shape);
+    std::shared_ptr<CylinderVolumeBounds> bounds =
+        std::make_shared<CylinderVolumeBounds>(tubs.getRMin(), tubs.getRMax(),
+                                               tubs.getZHalfLength(),
+                                               tubs.getDPhi() / 2);
+    Acts::Transform3 newTransform = Acts::Transform3::Identity();
+    newTransform.translate(transform.translation());
+    newTransform.rotate(transform.rotation() *
+                        Eigen::AngleAxisd(tubs.getSPhi() + 0.5 * tubs.getDPhi(),
+                                          Acts::Vector3::UnitZ()));
+    volumes.emplace_back(Experimental::DetectorVolumeFactory::construct(
+        portalGenerator, context, name, newTransform, bounds,
+        Experimental::tryAllPortalsAndSurfaces()));
+  } else if (shape.typeID() == GeoBox::getClassTypeID()) {
+    const GeoBox& box = static_cast<const GeoBox&>(shape);
+    std::shared_ptr<CuboidVolumeBounds> bounds =
+        std::make_shared<CuboidVolumeBounds>(
+            box.getXHalfLength(), box.getYHalfLength(), box.getZHalfLength());
+    volumes.emplace_back(Experimental::DetectorVolumeFactory::construct(
+        portalGenerator, context, name, transform, bounds,
+        Experimental::tryAllPortalsAndSurfaces()));
+  } else if (shape.typeID() == GeoSimplePolygonBrep::getClassTypeID()) {
+    const GeoSimplePolygonBrep& brep =
+        static_cast<const GeoSimplePolygonBrep&>(shape);
+    double xmin{0}, xmax{0}, ymin{0}, ymax{0}, zmin{0}, zmax{0};
+    brep.extent(xmin, ymin, zmin, xmax, ymax, zmax);
+    std::shared_ptr<CuboidVolumeBounds> bounds =
+        std::make_shared<CuboidVolumeBounds>(
+            (xmax - xmin) / 2, (ymax - ymin) / 2, (zmax - zmin) / 2);
+    volumes.emplace_back(Experimental::DetectorVolumeFactory::construct(
+        portalGenerator, context, name, transform, bounds,
+        Experimental::tryAllPortalsAndSurfaces()));
+  } else if (shape.typeID() == GeoTrd::getClassTypeID()) {
+    const GeoTrd& trd = static_cast<const GeoTrd&>(shape);
+    double x1 = trd.getXHalfLength1();
+    double x2 = trd.getXHalfLength2();
+    double y1 = trd.getYHalfLength1();
+    double y2 = trd.getYHalfLength2();
+    double z = trd.getZHalfLength();
+    if (y1 == y2) {
+      if (x1 <= x2) {
+        // y axis in ACTS is z axis in geomodel
+        std::shared_ptr<TrapezoidVolumeBounds> bounds =
+            std::make_shared<TrapezoidVolumeBounds>(x1, x2, z, y1);
+        auto rotationAngle = M_PI / 2;
+        Acts::Transform3 newTransform = Acts::Transform3::Identity();
```

Comment:
GeoModel has nice helper classes to compose the transforms. Maybe you want to take them to clean up the code

https://gitlab.cern.ch/GeoModelDev/GeoModel/-/blob/main/GeoModelCore/GeoModelKernel/GeoModelKernel/GeoDefinitions.h#L95-109